python/ray/data/tests/test_streaming_executor.py

```diff
@@ -0,0 +1,180 @@
+import pytest
+import asyncio
+import time
+from unittest.mock import MagicMock
+
+from typing import List, Any
+
+import ray
+from ray.data.context import DatasetContext
+from ray.data._internal.execution.interfaces import ExecutionOptions, RefBundle
+from ray.data._internal.execution.streaming_executor import StreamingExecutor
+from ray.data._internal.execution.streaming_executor_state import (
+    OpState,
+    build_streaming_topology,
+    process_completed_tasks,
+    select_operator_to_run,
+)
+from ray.data._internal.execution.operators.all_to_all_operator import AllToAllOperator
+from ray.data._internal.execution.operators.map_operator import MapOperator
+from ray.data._internal.execution.operators.input_data_buffer import InputDataBuffer
+from ray.data._internal.execution.util import make_ref_bundles
+
+
+@ray.remote
+def sleep():
+    time.sleep(999)
+
+
+def make_transform(block_fn):
+    def map_fn(block_iter):
+        for block in block_iter:
+            yield block_fn(block)
+
+    return map_fn
+
+
+def ref_bundles_to_list(bundles: List[RefBundle]) -> List[List[Any]]:
+    output = []
+    for bundle in bundles:
+        for block, _ in bundle.blocks:
+            output.append(ray.get(block))
+    return output
+
+
+def test_build_streaming_topology():
```

Comment:
Could we have a test covering the operator caching/common operator elimination/whatever we're calling it? Just to make sure that we're not doing redundant execution for operators that fan-out to multiple descendants.

Reply:
I ended up realizing our current implementation is completely nonsensical for forked operators (we'd need to support multiple output queues for that to make sense).

Added a test instead to raise an error if this case is detected.